pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -96,74 +118,184 @@ func (r *HTTPRouteReconciler) gapiToKumaRoute(ctx context.Context, mesh string,
+) error {
+	// First we list which existing objects are owned by this route.
+	// Then we iterate through the current refs and either create new ones or
+	// update existing ones. Finally, we delete any existing objects for refs
+	// which no longer exist.
+	ownerLabelValue := util_k8s.K8sNamespacedNameToCoreName(owner.Name, owner.Namespace)
+	labels := kube_client.MatchingLabels{
+		ownerLabel: ownerLabelValue,
+	}
+
+	list := &mesh_k8s.GatewayRouteList{}
+	if err := client.List(ctx, list, labels); err != nil {
+		return err
+	}
 
-		var filters []*mesh_proto.GatewayRoute_HttpRoute_Filter
+	type GatewayRef string
 
-		for _, filter := range rule.Filters {
-			kumaFilter, err := r.gapiToKumaFilter(ctx, mesh, namespace, filter)
-			if err != nil {
-				return nil, err
-			}
+	routesForGateway := map[GatewayRef]*mesh_k8s.GatewayRoute{}
 
-			filters = append(filters, kumaFilter)
+	for _, route := range list.Items {
+		selectorMatch := route.Spec.GetSelectors()[0].GetMatch()
+		// We know that GatewayRoutes owned by us contain exactly one selector,
+		// with the serviceTagForGateway format
+		tag := selectorMatch[mesh_proto.ServiceTag]
+
+		gatewayName, err := gatewayForServiceTag(tag)
+		if err != nil {
+			// TODO log and continue
+			return err
 		}
 
-		rules = append(rules, &mesh_proto.GatewayRoute_HttpRoute_Rule{
-			Matches:  matches,
-			Filters:  filters,
-			Backends: backends,
-		})
+		key := gatewayName.String()
+
+		// We might also be binding to a specific listener
+		if listener, ok := selectorMatch[mesh_proto.ListenerTag]; ok {
+			key = path.Join(key, listener)
+		}
+
+		// GatewayRoutes are cluster-scoped
+		routesForGateway[GatewayRef(key)] = &route
 	}
 
-	return &mesh_proto.GatewayRoute{
-		Selectors: selectors,
-		Conf: &mesh_proto.GatewayRoute_Conf{
-			Route: &mesh_proto.GatewayRoute_Conf_Http{
-				Http: &mesh_proto.GatewayRoute_HttpRoute{
-					Hostnames: hostnames,
-					Rules:     rules,
+	for gatewayRef, spec := range routeSpecs {
+		// First we need to figure out the key for this reference
+		gatewayNamespace := owner.Namespace
+		if ns := gatewayRef.Namespace; ns != nil {
+			gatewayNamespace = string(*ns)
+		}
+
+		gatewayKey := kube_types.NamespacedName{
+			Namespace: gatewayNamespace,
+			Name:      string(gatewayRef.Name),
+		}.String()
+
+		if listener := gatewayRef.SectionName; listener != nil {
+			gatewayKey = path.Join(gatewayKey, string(*listener))
+		}
+
+		// We either are already maintaining a GatewayRoute for this Ref
+		if route, ok := routesForGateway[GatewayRef(gatewayKey)]; ok {
+			route.Spec = spec
+
+			if err := client.Update(ctx, route); err != nil {
+				return errors.Wrapf(err, "could not update GatewayRoute for Gateway %s", gatewayKey)
+			}
+
+			delete(routesForGateway, GatewayRef(gatewayKey))
+			continue
+		}
+
+		// Or it's a new Ref
+		route := &mesh_k8s.GatewayRoute{
+			ObjectMeta: kube_meta.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", ownerLabelValue),
```

Comment:
I actually prefer using multiple separators, it's not like a file path with arbitrary levels. But I think since we can't use `/` a `.` may be somewhat surprising, changed everything to `-`.